test/middleware.ts

```diff
@@ -0,0 +1,264 @@
+        'another-header': 'another-header-value',
+        "method": "GET",
+        "path": "/hello/world",
+        "params": "?a=1&b=2",
+        "response-size": 12345,
+        "response-time": 99999,
+        "status-code": 200,
+        "x-forwarded-for": "foo",
+      })
+      assert.equal(masked, expected)
+      done()
+    })
+
+    var stream = createLineStream(function (line) {
+      cb(null, null, line)
+    })
+
+    var options = {
+      headers: ['some-header', 'another-header'],
+    };
+
+    var server = createServer(options, { stream: stream }, function (req, res, next) {
+      res.setHeader('content-length', 12345)
+      next()
+    })
+
+    request(server)
+    .get('/hello/world?a=1&b=2')
+    .set('some-header', 'some-header-value')
+    .set('another-header', 'another-header-value')
+    .set('x-forwarded-for', 'foo')
+    .expect(200, cb)
+  })
+
+  it('should allow logging from user-specified handlers', function (done) {
+    var cb = afterTest(2, function (err, res, line) {
+      if (err) return done(err)
+      var masked = line.replace(/response-time\":\d+/, 'response-time":99999')
+      expected = kayvee.format({
+        "global": 1,
+        "global2": 2,
+        "url": '/hello/world?a=1&b=2',
+        "method": "GET",
+        "path": "/hello/world",
+        "params": "?a=1&b=2",
+        "response-size": 12345,
+        "response-time": 99999,
+        "status-code": 200,
+        "x-forwarded-for": "foo",
+      })
+      assert.equal(masked, expected)
+      done()
+    })
+
+    var stream = createLineStream(function (line) {
+      cb(null, null, line)
+    })
+
+    var options = {
+      handlers: [
+        function(req, res) {return {"global": 1}},
+        function(req, res) {return {"global2": 2}},
+        function(req, res) {return {"url": req.url }},
+      ]
+    };
+
+    var server = createServer(options, { stream: stream }, function (req, res, next) {
+      res.setHeader('content-length', 12345)
+      next()
+    })
+
+    request(server)
+    .get('/hello/world?a=1&b=2')
+    .set('x-forwarded-for', 'foo')
+    .expect(200, cb)
+  })
+
+  it('should not log null or undefined values', function (done) {
+    var cb = afterTest(2, function (err, res, line) {
+      if (err) return done(err)
+      var masked = line.replace(/response-time\":\d+/, 'response-time":99999')
+      expected = kayvee.format({
+        "method": "GET",
+        "path": "/hello/world",
+        "params": "?a=1&b=2",
+        "response-size": 12345,
+        "response-time": 99999,
+        "status-code": 200,
+        "x-forwarded-for": "foo",
+      })
+      assert.equal(masked, expected)
+      done()
+    })
+
+    var stream = createLineStream(function (line) {
+      cb(null, null, line)
+    })
+
+    var options = {
+      // These values should not be logged
+      headers: ['this-header-dne'],
+      handlers: [
+        function() { return { "undefined": undefined } }
```

Comment:
It looks like we don't test null - should we add one for that as well? e.g. `"null": null`